packages/grid/_modules_/grid/hooks/features/core/useGridControlState.ts

```diff
@@ -19,49 +19,53 @@ export function useGridControlState(apiRef: GridApiRef) {
-        const hasSubStateChanged = oldState !== newSubState;
 
-        if (updatedStateIds.length >= 1 && hasSubStateChanged) {
-          // Each hook modify its own state and it should not leak
-          // Events are here to forward to other hooks and apply changes.
-          // You are trying to update several states in a no isolated way.
-          throw new Error(
-            `You're not allowed to update several sub-state in one transaction. You already updated ${updatedStateIds[0]}, therefore, you're not allowed to update ${controlState.stateId} in the same transaction.`,
-          );
+        // TODO newSubState !== controlState.propModel shouldn't be necessary
+        // but we need it for the initial state.
+        if (newSubState !== oldState && newSubState !== controlState.propModel) {
+          updatedStateIds.push(controlState.stateId);
         }
 
-        if (hasSubStateChanged) {
-          if (controlState.propOnChange) {
-            const newModel = newSubState;
-            if (controlState.propModel !== newModel) {
-              controlState.propOnChange(newModel);
-            }
-            shouldUpdate =
-              controlState.propModel === undefined || controlState.propModel === newModel;
-          } else if (controlState.propModel !== undefined) {
-            shouldUpdate = oldState !== controlState.propModel;
-          }
-          if (shouldUpdate) {
-            updatedStateIds.push(controlState.stateId);
-          }
+        // The state is controlled, the prop should always win
+        if (controlState.propModel !== undefined && newSubState !== controlState.propModel) {
+          ignoreSetState = true;
         }
       });
 
+      if (updatedStateIds.length > 1) {
+        // Each hook modify its own state and it should not leak
+        // Events are here to forward to other hooks and apply changes.
+        // You are trying to update several states in a no isolated way.
+        throw new Error(
+          `You're not allowed to update several sub-state in one transaction. You already updated ${
+            updatedStateIds[0]
+          }, therefore, you're not allowed to update ${updatedStateIds.join(
+            ', ',
+          )} in the same transaction.`,
+        );
+      }
+
       return {
-        shouldUpdate,
+        ignoreSetState,
         postUpdate: () => {
           updatedStateIds.forEach((stateId) => {
-            if (controlStateMap[stateId].onChangeCallback) {
-              const model = controlStateMap[stateId].stateSelector(newState);
-              controlStateMap[stateId].onChangeCallback!(model);
+            const controlState = controlStateMap[stateId];
+            const model = controlStateMap[stateId].stateSelector(newState);
+
+            if (controlState.propOnChange) {
+              controlState.propOnChange(model);
+            }
+
+            if (controlState.onChangeCallback) {
+              controlState.onChangeCallback!(model);
```

Comment:
I was thinking if shouldn't we add a `modelChangeEvent` to `ControlStateItem` and publish automatically the event. This callback in 100% of the cases is only used to publish the event.

Reply:
Ok done